### ISSUE TITLE
Large number of "In Progress" records from previous years

### ISSUE THREAD

**Issue by kitenetter** [BiologicalRecordsCentre/iRecord#670]
This arises because I was puzzled by the apparent drop in the number of records in 2014:

![image](https://user-images.githubusercontent.com/16039111/64634567-6f580b00-d3f5-11e9-8a9d-c2cb47fc598c.png)

On investigation this results from a very large number of records, in both 2014 and 2015, being given a record_status of "I" (= In progress):

![image](https://user-images.githubusercontent.com/16039111/64634781-f6a57e80-d3f5-11e9-8c94-634eca0a32a9.png)

I don't understand why we have so many records "In progress". They originate from a variety of surveys:

![image](https://user-images.githubusercontent.com/16039111/64636660-13dc4c00-d3fa-11e9-9451-340cb66f7285.png)

Some of these surveys look as if they contain records that should not be made available, e.g. "CWS digitisation species name check" (linked to Norfolk Biodiversity Information Service) is missing vital information such as recorder name; and the "(inactive) Old Paths Survey" may also be deliberately hidden. But others seem to be genuine records, and in any case my understanding is that the "I" status is not intended to flag up records that are not to be verified or shared.

It's not clear to me whether the "In progress" records are in fact being displayed to verifiers - on the face of it it would seem not, since so many of them for 2014 and 2015 have not been verified. However, if I look for one of the "In progress" records from the "CWS digitisation species name check" that is of a bee-fly (record id 1961906) it is visible to me on the new verification page, but only if I filter for "All records" instead of "Pending". (I tried to check on the old verification as well but at the moment that gets stuck if I try to set it to "All records".)

So maybe these "I" records are visible to verifiers, but have not been processed because they are not included in the "Pending" view. Arguably these records should be reset to "C" (= not reviewed) so that they can be processed in the normal way. However, before going down that route we need to know more about why all these records are given status "I", and whether they are in fact genuine records or not. If they are then they ought to be released for verification, but before doing that I would want to do a taxonomic analysis to review what this would mean for verifiers' workloads.

Can @johnvanbreda or @DavidRoy shed any light on this?

**Comment by johnvanbreda** [comment]
Record_status=I (in progress) was intended to be used where a data entry form had been partially filled in and submitted, but the records are in a state where there is more work to do on them before they are complete and ready to review. E.g. you want to add a photo to the record. As far as I know it has rarely been used for any genuine reasons but as you say there may be a few. If records are complete from the data entry perspective but should be withheld from verification and other uses for some reason, then the release_status flag can be set to 'U'.

As the default verification filter is record_status=C, the in progress records are hidden from the default view but can be found under the All records view.

I agree that most, if not all of these should be set to 'C' at some point.

**Comment by kitenetter** [comment]
The first requirement is to resolve #486 

Once that is done I propose to start reassigning these records from "I" to "C".

Question for @johnvanbreda : if the UKBMS and NPMS records are updated from "I" to "C" will they automatically be auto-verified, or is another operation required to force the auto-verification process?

**Comment by johnvanbreda** [comment]
If you change occurrences.record_status and updated_on via the database, you should also insert work_queue entries to ensure the cache builder table is updated. See https://indicia-docs.readthedocs.io/en/latest/administrating/warehouse/scheduled-tasks.html#work-queue.

Once the cache builder update occurs the auto_verify module should pick the records up.

**Comment by kitenetter** [comment]
Currently awaiting responses from NBIS re "CWS digitisation species name check" and from NPMS re "(inactive) Old Paths Survey".

Hold off on action until the forthcoming verifiers' newsletter has been sent out, which will forewarn verifiers of this additional influx of records (mostly butterflies, hopefully most of which will be dealt with via auto-verification once they are made available).

**Comment by burkmarr** [comment]
@kitenetter - what criteria are we going to use to determine which records should be changed from status of 'I' to 'C'? Is it just going to be all UKBMS and NPMS records?

**Comment by kitenetter** [comment]
@burkmarr this is going to take a while as there are different issues for the different surveys. Let me know if you want me to set these up as separate issues for each survey we tackle.

The first one is ready to go. There are 31,312 records involved, from survey id 292 (CWS digitisation species name check). In this case all the records are to be marked as deleted. The records can be found via:

`SELECT o.id, o.deleted
FROM occurrences o
JOIN samples s ON o.sample_id = s.id
WHERE s.survey_id = 292 AND o.deleted = false`

Each of those occurrences needs to have o.deleted set to 'true'.

As far as I'm aware there are no other actions required, unless anything needs to be done to enforce the cache or elastic versions to update.

**Comment by burkmarr** [comment]
@johnvanbreda. I've done the update on the occurrences table. I think I need to add tasks to the work_queue to rebuild the cache (task_cache_builder_update) - is that right? If so, can I just get away with inserting a single task identifying the survey to which all these records belong, or do I need to insert one for each record?

**Comment by johnvanbreda** [comment]
You can either insert one per record for a proper refresh of the cache data, or if you prefer, write a pgAdmin SQL script that updates the contents of cache_occurrences_functional for the relevant records (setting record_status and record_substatus).

**Comment by burkmarr** [comment]
Hi John - this first lot was just flagging as deletions - 31,000+ of them. Shall I just delete the corresponding records from the two cach_occurrences table? Seems better than adding all those jobs to the work queue.

**Comment by kitenetter** [comment]
The second dataset is also for deletion. 
- Survey ID = 229 ("(inactive) Old Paths survey")
- Number of records to delete = 11,926 (this includes a very few records that have been verified, but the decision is to delete the entire dataset)

The records can be found via:

`SELECT o.id, o.deleted
FROM occurrences o
JOIN samples s ON o.sample_id = s.id
WHERE s.survey_id = 229 AND o.deleted = false`

Each of those occurrences needs to have o.deleted set to 'true'.

(If the decision is to delete direct from the cache tables then that will need doing as well of course.)

**Comment by johnvanbreda** [comment]
@burkmarr yes, just deleting from cache_occurrences_* is fine as long as you are confident in the SQL.

**Comment by burkmarr** [comment]
@kitenetter - the first two (surveys 'CWS digitisation species name check' and '(inactive) Old Paths survey') are done.

**Comment by kitenetter** [comment]
Not yet found time to address this further.

**Comment by kitenetter** [comment]
This is the current position for all records with verification status = "In progress":

- UKBMS (website_id 27) with 558422 records - to be actioned following communication to verifiers in autumn 2022
- Lincolnshire Environmental Records Centre (id 45) with 8508 records - no iRecord sharing agreement, no action needed
Consultants Portal (id 44) with 5012 records - these are records of vertebrates and most seem to be of protected species, e.g. bats; not sure if there is a good reason for these to be marked as "In progress"
- iRecord (id 23) with 841 records - two test records to delete; all the others are from the "Moors for the Future" bumblebee transects, which BWARS do not wish to make use of, so can stay as they are
- Norfolk Biodiversity Information Service (id 7) with 160 records - all are butterfly records from a local implementation of a transect survey, need to check whether MBIS want these to be available for verification
- SpotABee (id 86) with 17 records - these are not species records and can be ignored
- GIGL (id 51) with 13 records - no obvious reason why these should not be mobilised, will ask GIGL
- ERCCIS (id 31) with 1 record - no iRecord sharing, can be ignored

Actions:
1. MH to liaise with BC to inform county verifiers before proceeding to update UKBMS records in autumn 2022
2. MH to ask NBIS and GIGL if their records can be mobilised

Question for @johnvanbreda - is there a legitimate reason why consultant portal data might need to remain as "In progress" (e.g. is this linked to the embargo process)?

**Comment by kitenetter** [comment]
@burkmarr I'm finally getting back to this issue, and would like to tackle the large number of UKBMS records next. Are you able to put some time (hopefully a short amount!) into this over the next week or two? Before you take any action I need to contact the butterfly verifiers, so if you can let me know when you will be able to work on this I will contact them a few days in advance.

Because this is such a large update we might need to think about when is best to run it.

- UKBMS (website_id 27) has 558422 records that currently have record_status=I
- All these records need to be set to record_status=C
- Do we need to update the updated_on date? Arguably we don't, since we are not making any change to the record itself, we are just correcting an error in the status.
- John has previously said "If you change occurrences.record_status [and updated_on] via the database, you should also insert work_queue entries to ensure the cache builder table is updated. See https://indicia-docs.readthedocs.io/en/latest/administrating/warehouse/scheduled-tasks.html#work-queue.
- We also need to ensure that following the status change the records are all passed through auto verification. John has said "Once the cache builder update occurs the auto_verify module should pick the records up".

@johnvanbreda is there any reason why we need to change the updated_on date?

**Comment by burkmarr** [comment]
@kitenetter - am I right in thinking that not updating the `updated_on` field will avoid records being flagged for reverification? As far as I am aware there is no compelling reason to update it in this case, but @johnvanbreda may know better?

I'm wondering if we should aim to do this update in batches to manage the work queue workload?

**Comment by kitenetter** [comment]
@burkmarr we do want the records to be verified, but I think that will be achieved by setting the status to "C", i.e. "not reviewed" (and then ensuring the auto-verification kicks in). So I'm not aware of any need to change `updated_on` but it would good to have confirmation from John.

Batches sound like a good idea.

**Comment by johnvanbreda** [comment]
@kitenetter 

> Question for @johnvanbreda - is there a legitimate reason why consultant portal data might need to remain as "In progress" (e.g. is this linked to the embargo process)?

Yes - the Consultant Portal projects were sometimes setup so that records were not going to be released for wider consumption till the project finished. Given that this website has been discontinued we'd need to try and trace the project coordinators (consultants?) to see if it is OK to release the records.

The re-verification is not triggered by an alteration to the updated_on field. Re-verification only happens when a record is altered via the warehouse's ORM code (i.e. not a backdoor direct update using SQL) since the business logic is in the model code. In fact the business logic actually specifies that a change to release status alone is not going to re-trigger verification.

Updating updated_on will have the effect of causing the data cleaner rules to be re-triggered though.

**Comment by kitenetter** [comment]
Thanks John. No action required for the Consultants Portal at the moment.

For the UKBMS records it sounds like it is desirable to update the `updated_on` attribute so as to apply the current record cleaner rules.

Based on the latest info from John I think what we are now proposing to do is:
- For the UKBMS (website_id 27) there are 558422 records that currently have record_status=I - these will be set to record_status=C
- Edit the updated_on date to the date of this update
- insert work_queue entries to ensure the cache builder table is updated

@johnvanbreda is that the correct sequence, and will that result in all these things happening:
- Records will have current record cleaner rules applied
- Auto verification will be applied to all these records
- Any records that do not pass the auto-verification will be available as pending records for the relevant verifiers

@burkmarr if that is the correct sequence we can proceed. I will email the butterfly verifiers today to warn them that these records will be appearing over the coming weeks, and then we can start the process any time next week (in batches as you suggested).

**Comment by johnvanbreda** [comment]
@kitenetter that sequence is correct. I would suggest doing the first 2 steps in a single query so they are done in unison.

**Comment by burkmarr** [comment]
@kitenetter - okay that's fine with me.

**Comment by burkmarr** [comment]
@kitenetter - the last batch of UKBMS records finished going through yesterday evening.

**Comment by kitenetter** [comment]
Thanks @burkmarr good to get all those done.

This issue is nearly completed, there are just two loose ends for me to follow up, affecting small numbers of records:
- Norfolk Biodiversity Information Service (id 7) with 160 records - all are butterfly records from a local implementation of a transect survey, need to check whether NBIS want these to be available for verification
- GIGL (id 51) with 13 records - no obvious reason why these should not be mobilised, will ask GIGL

**Comment by kitenetter** [comment]
@burkmarr it appears that a proportion of the butterfly records have not had the record cleaner rule checks applied, and thus are not being autoverified. I think it is quite a small proportion but I can't say precisely because I don't have any way of filtering for records that haven't been rule-checked. There are definitely some though!

Is it possible to re-trigger the record cleaner process for the ones that haven't been picked up?

**Comment by burkmarr** [comment]
@kitenetter - I don't understand how that can have happened since I'm sure that I updated the `updated_on` field on all of them which should have triggered the data cleaner rules. I can do it again if we can identify the problem records but, like you, I don't know how to identify them. Any thoughts @johnvanbreda?

**Comment by burkmarr** [comment]
@johnvanbreda - the following query returns no records. Would record cleaner rule checks have been applied to these records when their status was 'I'?
```
select * from cache_occurrences_functional
where website_id=27 and applied_verification_rule_types = null
```

**Comment by burkmarr** [comment]
@kitenetter - can you give me a few examples of records which you know were updated from record status 'I' to 'C' but have not had the record cleaner rules applied since then?

**Comment by kitenetter** [comment]
@burkmarr some examples:

**Oddities that I think are NOT related to the recent "In progress" updates** (I think this might be a separate issue)
- Occurrence ID 18279533: the verification page shows no auto-check pass or fail symbols; the download says "passed checks"; it has not been auto-verified (I would expect it pass the checks)
- 18279532: same as above

**Oddities that ARE related to the recent "In progress" updates**
- 2583492: the verification page shows the symbol for "Record not yet checked against rules"; the download is null for the auto checks; it has not been auto-verified (I would expect it pass the checks)
- 2583067: same as above
- 1493135: same as above
- 1483817: same as above

There are lots more examples if needed!

**Comment by burkmarr** [comment]
Re my comment https://github.com/BiologicalRecordsCentre/iRecord/issues/670#issuecomment-1721274073, I've realised that `applied_verification_rule_types = null` should have been `applied_verification_rule_types is null`. Fixing the query I find that there are over 4 million records from website 27 where no auto-verification rules have been run, but I suppose that this includes many historic records. The problem is identifying those records whose status we can changed from 'I' to 'C'. I think that the best I can do is to limit the selection to those records updated on or after the date when I started doing the change.

The following query identifies 208,827 records:
```SQL
select count(*)
from cache_occurrences_functional
where website_id=27
and applied_verification_rule_types is null
and updated_on >= '2023-09-01'
```

@johnvanbreda and @kitenetter - I can update the `updated_on` (and `updated_by`) fields on the `occurrences` table for those records to trigger the cleaner rules and add work_queue records to rebuild the cache table entries. I can do it in batches of 50,000 again if that sounds okay to you two?

( I'm still none the wiser as to why that workflow didn't work on all the records before though.)

**Comment by kitenetter** [comment]
@burkmarr from my perspective I agree that it would be good to re-apply the rule checks and auto-verification to the records from the date of the change.

**Comment by burkmarr** [comment]
@kitenetter - hopefully most (if not all) of these should be sorted out now - those 209,000 records should have all had the data cleaner rule checks applied again now.

**Comment by kitenetter** [comment]
@kitenetter to review whether any further work is required here (I think we have addressed all the main issues at least).

**Comment by kitenetter** [comment]
All relevant issues addressed - closing.
